metapool/sequencers.py

```diff
@@ -141,15 +141,77 @@ def _get_machine_code(instrument_model):
         If the instrument model is malformed and does not contain a valid
         machine code.
     """
-    # the machine code represents the first 1 to 2 letters of the
-    # instrument model
-    machine_code = re.compile(r'^([a-zA-Z]{1,2})')
-    matches = re.search(machine_code, instrument_model)
-    if matches is None:
-        raise ValueError('Cannot find a machine code. This instrument '
-                         'model is malformed %s. The machine code is a '
-                         'one or two character prefix.' % instrument_model)
-    return matches[0]
+    # the machine code is everything before the first number
+    # (we expect these to be letters, so pattern could be improved ...)
+    matches = re.match(r"^(.*?)\d.*", instrument_model)
+
+    if matches is not None:
+        machine_code = matches.group(1)
+        if machine_code != "":
+            return machine_code
+
+    raise ValueError(f"Cannot find a machine code; the instrument "
+                     f"model '{instrument_model}' is malformed.")
+
+
+def _get_model_by_machine_prefix(instrument_prefix, sequencer_types=None):
+    """Get the instrument model by its machine prefix.
+
+    Parameters
+    ----------
+    instrument_prefix: str
+        The machine prefix of the instrument, e.g., 'MN' for MN01225.
+    sequencer_types: MappingProxyType, optional
+        A mapping of available sequencer types. If None, the
+        sequencer types will be loaded from the YAML file.
+
+    Returns
+    -------
+    MappingProxyType
+        Immutable dictionary of the instrument model details.
+
+    Raises
+    ------
+    ValueError
+        If the instrument prefix is not recognized or if multiple
+        sequencer types match the given prefix.
+    """
+    models_w_prefix = get_sequencers_w_key_value(
+        _MACHINE_PREFIX_KEY, instrument_prefix,
+        existing_types=sequencer_types)
+    if len(models_w_prefix) == 0:
+        raise ValueError(
+            f"Unrecognized {_MACHINE_PREFIX_KEY} '{instrument_prefix}'.")
+    elif len(models_w_prefix) > 1:
+        raise ValueError(
+            f"Found {len(models_w_prefix)} sequencer types with "
+            f"{_MACHINE_PREFIX_KEY} '{instrument_prefix}': "
+            f"{', '.join(models_w_prefix)}.")
+    # end if got an unexpected number of sequencer types w given prefix
+
+    inst_model_type = next(iter(models_w_prefix))
+    instrument_model = _get_model_by_sequencer_type_name(
+        inst_model_type, sequencer_types=models_w_prefix)
+    return instrument_model
+
+
+def get_model_by_instrument_id(instrument_id, sequencer_types=None):
```

Comment:
Is `sequencer_types` used for anything else than the tests? If no, why not create a `sequencer_types_multiple.yml` with the multiple errors?

Reply:
It does get used in places other than the tests: `get_model_by_instrument_id` is called inside `get_model_and_center`, which already has a sequencer_types object, so it passes it to `get_model_by_instrument_id` instead of reloading it from disk.